### DISCUSSION TITLE
コールバックがバグの温床になっているので解決したい...(after_createなど)

### DISCUSSION THREAD

**Discussion by hochan1921** [roppongirb/discussions#44] [General]
アプリケーションが大きくなってきて動作検証の考慮事項が多くなってきました。

社内で議論したのですが、1つの要因としてあがったが、コールバックです。
after_create, after_updateなど、裏側で処理が走るものについて考慮できずに実装してしまうケースが多いです。

**Comment by hatappi** [comment]
## Callback を使っているユースケース
- 社内の DB に変更があった時に外部のサービスに対しても同期をかけたい

## 起きた問題
- 規模があがるにつれてモデルに紐づく callback が増えてきた
- callback 起因で起きた問題が発見しずらい？
- dev, test とはデータが異なるのでデータの更新がトリガーになるようなものだとテストで発見できないことも

## 話したいこと
- callback をどう扱っている

**Comment by hatappi** [comment]
## 他のユースケース
- before_create/update でデフォルト値
  - before_create が呼ばれるのは新規にレコードが追加される直前に呼ばれる callback 
  - DB のカラム定義時にdefault値を設定する？
- after_create は使ってる部分はあるけど after_update までは使っていない
- 完全に同期する動作を保証したい時は使う (User 消したら他のやつも絶対消す！！！)
  - 影響範囲が明確な時など

## 困ったケース
- model 同士が 1対1となっていて、片方の model がもう片方の model のカラムを更新するが更新する側が外部キーをもってないので callback 時に model があることが必須になっているとエラー
- after_create で他のモデルを更新する時に、他のモデルが更新されない時に気づく

https://railsguides.jp/active_record_callbacks.html#%E5%88%A9%E7%94%A8%E5%8F%AF%E8%83%BD%E3%81%AA%E3%82%B3%E3%83%BC%E3%83%AB%E3%83%90%E3%83%83%E3%82%AF

**Comment by hatappi** [comment]
## callback 起因で起きた問題が発見しずらい？
- callback が呼ばれている時点では問題になってなくて後々問題が明るみになる
   - 影響範囲が見えづらい
   - 意図しないところが更新されている ==> callback .... みたいな流れ
- コードリーディングの観点からパッとモデルをみた時に callback がいくつもあると何が起きるのか追いづらくレビューも苦労。。。

**Comment by hatappi** [comment]
## callback を良い感じに使うもしくは別の方法で目的を達成する
- callback ではなくモデルの更新をトリガーにして処理をする
- **service 層などまとまった部分で一連のモデルの更新処理を定義する**

```rb
class RegisterService
  def create_new_user
     ActiveRecord::Transcation do 
       user = User.create(...)
       chat = Chat.create(user: user)
    end
  end
end

~~~

class RegisterController
   def new
      RegisterSerview.create_new_user(params)
   end
end

```

Service におくか Model におくかは #29 を見てください！！

**Reply by hatappi** [reply]
**after_create を使うにしても rescue とかでエラーをひろって、問題がおきた時に Slack 通知にするなどして気付けるようにする。**
